### PR TITLE
Geoms (ucrGeom) + dotplot + sdgLayerOptions + ucrGeomListWithAes

### PR DESCRIPTION
**Pull request by FrancoisJRenaud**
Will be editing geoms a little until we get started with the linking construction site which I know believe to be urgent...
This pull request is in progress...

### CODE REVIEW

File: instat/ucrGeom.vb [hunk 1]
```diff
@@ -428,25 +430,49 @@ Public Class ucrGeom
...
-        clsgeom_dotplot.AddLayerParameter("origin", "numeric", "0", lstParameterStrings:={1})
-        clsgeom_dotplot.AddLayerParameter("right", "Boolean", "TRUE")
-        clsgeom_dotplot.AddLayerParameter("width", "numeric", "0.9", lstParameterStrings:={1})
-        clsgeom_dotplot.AddLayerParameter("drop", "Boolean", "FALSE")
+
+        'Aesthetics parameters
+        clsgeom_dotplot.AddAesParameter("x", strIncludedDataTypes:={"factor", "numeric"}) 'x is partially mandatory, need to be given value "" when not filled.
+        clsgeom_dotplot.AddAesParameter("y", strIncludedDataTypes:={"numeric", "factor"})
+        'When the binaxis is y, then if x is a factor or character, several dotplots lines are displayed according to each group along the x axis. 
+        'When the binaxis is x, and y is a factor or character, dot stacks are split according to the y mapping, but not aligned with the labels along the y axis. So all stacks stay at the bottom. Also this is visible only if stackratio < 1, and stackgroups = FALSE.
```
Comment (FrancoisJRenaud) [inline instat/ucrGeom.vb]: You may want to try 
> ggplot(x) + geom_dotplot( mapping = aes(x=manufacturer, y=drv), colour = "red", method = "histodot", binwidth = 1, stackratio = 0.1, dotsize = 0.2 )
> ggplot(x) + geom_dotplot( mapping = aes(x=manufacturer, y=drv), colour = "red", method = "histodot", binwidth = 1, stackratio = 0.1, dotsize = 0.2, stackgroups = TRUE )
> ggplot(x) + geom_dotplot( mapping = aes(x=manufacturer, y=drv), colour = "red", method = "histodot", binwidth = 1, stackratio = 0.2, dotsize = 0.2 )
> ggplot(x) + geom_dotplot( mapping = aes(x=manufacturer, y=drv), colour = "red", method = "histodot", binwidth = 1, stackratio = 1, dotsize = 0.2 )
> ggplot(x) + geom_dotplot( mapping = aes(x=manufacturer, y=drv), colour = "red", method = "histodot", binwidth = 1, stackratio = 2, dotsize = 0.2 )
> ggplot(x) + geom_dotplot( mapping = aes(x=manufacturer, y=drv), colour = "red", method = "histodot", binwidth = 1, stackratio = 0.1, dotsize = 0.2 )

**Comment by dannyparsons** [comment]
Is the geom complete and tested in the dialogs too?

**Comment by FrancoisJRenaud** [comment]
eu, not really yet. Today was pretty inefficient. Will be doing this tomorrow fully when also editing the dialogue.

**Comment by FrancoisJRenaud** [comment]
The thing is that each time I try to test I get a bug because of the links... That's why I previously stopped testing thinking I would first work on these bug, then it appeared to be a time soaking task and I was told to not focus exclusively on it. Now would be good to start tackling this linking soon, so that the graphics can be cleaned from these myriads of obstacles...

Now about this pull request, I ll first ask @maxwellfundi  to go through it when I ve made something clean as discussed during last team skype. Don't worry I ll keep it reasonably short.

**Comment by dannyparsons** [comment]
Yes, you should be able to work on geoms separately from the linking issue. Maybe doing it through general and always just do a single layer is a good way to test a geom definition. Testing its specific dialog is different  but that shouldn't be too problematic, even with the linking issue.

**Comment by dannyparsons** [comment]
I'm nervous about the title of your pull request - it looks like you are going to edit a lot of things. Working on the geom definition should be a single pull request. Another one for the dotplot dialog and other pull requests for user controls or sub dialogs etc.

**Comment by FrancoisJRenaud** [comment]
Mhm, yeah, sorry. In the end I didn't work on the dotplot dialogue or on sdgLayerParameters or on ucrGeomListWithAes. The thing is that I thought I couldn't test what I did properly without also making the dlgDotPLot work. Then I did what you said, whent to generalForGraphics instead. but then got caught by the bugs there too. I was exploring the bugs for the LayerParameters as they were avoiding me from making progress. I found something after hours, got excited and solved. 

Should have done on another pull request. But then eople didn't seem like they were working on these files... Mhm, I edited RFunction though which is probably not very smart. Hopefully things are ok. 

So in summary, just did the geom edit and the bug. That's already one too much, but I ll stop here promised. Will have someone going through before you have a look @dannyparsons.

File: instat/ucrGeom.vb [hunk 2]
```diff
@@ -432,16 +435,16 @@ Public Class ucrGeom
         clsgeom_dotplot.strGeomName = "geom_dotplot"
 
         'Aesthetics parameters
-        clsgeom_dotplot.AddAesParameter("x", strIncludedDataTypes:={"factor", "numeric"}) 'x is partially mandatory, need to be given value "" when not filled.
-        clsgeom_dotplot.AddAesParameter("y", strIncludedDataTypes:={"numeric", "factor"})
+        clsgeom_dotplot.AddAesParameter("x", strIncludedDataTypes:={"numeric", "factor", "character"}) 'x is partially mandatory, need to be given value "" (or anything) when not filled.
```
Comment (FrancoisJRenaud) [inline instat/ucrGeom.vb]: Tried to add characters in dotplot aes. Maybe still needs a bit of testing/discussions, but works fine in R so ... If not happy, will delete. Still about #2239

File: instat/ucrGeom.vb [hunk 3]
```diff
@@ -465,11 +468,11 @@ Public Class ucrGeom
         'Warning/question: as bins with zero counts have zero dots, what is meant by remove ? Can't see any difference on the graphs when changing drop.
         'Global Layer parameters
         'Warning: stat is not a parameter of geom_dotplot. Anyway the stat that is used is "bindot".
-        clsgeom_bar.AddLayerParameter("show.legend", "boolean", "TRUE")
-        clsgeom_bar.AddLayerParameter("position", "list", Chr(34) & "stack" & Chr(34), lstParameterStrings:={Chr(34) & "stack" & Chr(34), Chr(34) & "dodge" & Chr(34), Chr(34) & "identity" & Chr(34), Chr(34) & "jitter" & Chr(34), Chr(34) & "fill" & Chr(34)})
+        clsgeom_dotplot.AddLayerParameter("show.legend", "boolean", "TRUE")
```
Comment (FrancoisJRenaud) [inline instat/ucrGeom.vb]: just a few typos.

File: instat/ucrGeom.vb [hunk 4]
```diff
@@ -935,7 +938,9 @@ Public Class ucrGeom
     End Sub
     Public Event GeomChanged(sender As Object, e As EventArgs)
     Private Sub cboGeomList_SelectedIndexChanged(sender As Object, e As EventArgs) Handles cboGeomList.SelectedIndexChanged
+        'Here, the clsCurrGeom is chosen according to the selected geom in cboGeomList. Then the Rcommand of the GeomFunction is chosen accordingly.
         clsCurrGeom = lstAllGeoms(cboGeomList.SelectedIndex)
+        'Warning/question/task : would be nice to erase parameters in the clsGeomFunction when geom has indeed changed... Maybe clear parameters in RFunction when Rcommand is actually changed (really changed)...
```
Comment (FrancoisJRenaud) [inline instat/ucrGeom.vb]: See RFunction edit above. I can delete this depending on what has been discussed.

File: instat/ucrLayerParameters.Designer.vb [hunk 5]
```diff
@@ -32,6 +32,10 @@ Partial Class ucrLayerParameters
         Me.UcrLayerParamsControls11 = New instat.ucrLayerParamsControls()
```
Comment (FrancoisJRenaud) [inline instat/ucrLayerParameters.Designer.vb]: Added extra ucrLayerParameterControls as there were 17 layerParameters in geom_dotplot. Hopefully won't have to change anymore. There are a lot of parameters because some of them are only relevant when others are in some setting...
Also had to adapt size here

File: instat/ucrLayerParameters.vb [hunk 6]
```diff
@@ -27,7 +27,7 @@ Public Class ucrLayerParameters
         InitializeComponent()
 
         ' Add any initialization after the InitializeComponent() call.
-        lstLayerParameterControl.AddRange({UcrLayerParamsControls1, UcrLayerParamsControls2, UcrLayerParamsControls3, UcrLayerParamsControls4, UcrLayerParamsControls5, UcrLayerParamsControls6, UcrLayerParamsControls7, UcrLayerParamsControls8, UcrLayerParamsControls9, UcrLayerParamsControls10, UcrLayerParamsControls11, UcrLayerParamsControls12, UcrLayerParamsControls13})
+        lstLayerParameterControl.AddRange({UcrLayerParamsControls1, UcrLayerParamsControls2, UcrLayerParamsControls3, UcrLayerParamsControls4, UcrLayerParamsControls5, UcrLayerParamsControls6, UcrLayerParamsControls7, UcrLayerParamsControls8, UcrLayerParamsControls9, UcrLayerParamsControls10, UcrLayerParamsControls11, UcrLayerParamsControls12, UcrLayerParamsControls13, UcrLayerParamsControls14, UcrLayerParamsControls15, UcrLayerParamsControls16, UcrLayerParamsControls17})
```
Comment (FrancoisJRenaud) [inline instat/ucrLayerParameters.vb]: Adding the new ucrLayerParamControls

File: instat/ucrLayerParameters.vb [hunk 7]
```diff
@@ -77,7 +77,7 @@ Public Class ucrLayerParameters
...
+    Private Sub UcrLayerParamsControls_RParameterChanged(ucrControl As ucrLayerParamsControls) Handles UcrLayerParamsControls1.RParameterChanged, UcrLayerParamsControls2.RParameterChanged, UcrLayerParamsControls3.RParameterChanged, UcrLayerParamsControls4.RParameterChanged, UcrLayerParamsControls5.RParameterChanged, UcrLayerParamsControls6.RParameterChanged, UcrLayerParamsControls7.RParameterChanged, UcrLayerParamsControls8.RParameterChanged, UcrLayerParamsControls9.RParameterChanged, UcrLayerParamsControls10.RParameterChanged, UcrLayerParamsControls11.RParameterChanged, UcrLayerParamsControls12.RParameterChanged, UcrLayerParamsControls14.RParameterChanged, UcrLayerParamsControls15.RParameterChanged, UcrLayerParamsControls16.RParameterChanged, UcrLayerParamsControls17.RParameterChanged
```
Comment (FrancoisJRenaud) [inline instat/ucrLayerParameters.vb]: Could probably delete this sub instead ?

File: instat/ucrLayerParamsControls.vb [hunk 8]
```diff
@@ -17,7 +17,7 @@
 
 Public Class ucrLayerParamsControls
     Public clsGeomFunction As RFunction
-
+    Private bEditGeomFunction As Boolean = True 'Warning/question: is there no load sub for this ucr ?     We usually avoid to set defaults like this.
```
Comment (FrancoisJRenaud) [inline instat/ucrLayerParamsControls.vb]: Adding this to be able to setup the layerParameters values and others without editing the geomFunction, which was the cause of all the LayerParameter filling and unfilling bugs... Similar method is found in ucrGeomListWithAes. 

Then I was wondering if I should add a load function to set the default value of bEditGeomFunction ?

**Comment by dannyparsons** [comment]
Great, we've made progress. Yes please stop here, and remember to be restrictive on what you do in a single branch - I think it will actually help your organisation of work too. You can always have multiple branches if you manage it carefully. 
The issue isn't so much about others working on the code, it's about the complexity of having lots of long comments threads on a branch, that become really time consuming for me to manage and keep track of.
It also prevents complete tasks being added into the code because everything on the branch must be complete before anything is accepted.

File: instat/ucrLayerParamsControls.vb [hunk 9]
```diff
@@ -26,15 +26,18 @@ Public Class ucrLayerParamsControls
 
     Public Sub SetGeomFunction(clsTempFunction As RFunction)
         clsGeomFunction = clsTempFunction
-        RaiseEvent RParameterChanged(Me)
+        'RaiseEvent RParameterChanged(Me) 'Warning : this was source of problems... the controls should adapt to the content of clsGeomFunction first, then when controls are changed, clsGeomFunction is changed in turn according to the content of the controls.
```
Comment (FrancoisJRenaud) [inline instat/ucrLayerParamsControls.vb]: This was the first incoherent thing done: changing the geom function according to the content of the ucrLayerParamControls when the geomFunction has been changed ...

File: instat/clsRFunction.vb [hunk 10]
```diff
@@ -34,9 +34,13 @@ Public Class RFunction
...
     Public Event ParametersChanged()
 
-    Public Sub SetRCommand(strTemp As String)
+    Public Sub SetRCommand(strTemp As String)
+        'Warning/Question/Task: would be nice to clear the RParameters when the RCommand is changed (really changed)... Don't know if this would have any consequence elsewhere... Could comment it out if needed.
+        If strRCommand <> strTemp Then
+            ClearParameters()
         strRCommand = strTemp
-        bIsAssigned = False
+        End If
```
Comment (dannyparsons) [inline instat/clsRFunction.vb]: I can see why this might be sensible, but I can think of dialogs where this will break them. I'm happy that if you want to clear parameters you have to do that manually on a separate line. It makes the dialog code more understandable too.

File: instat/ucrLayerParamsControls.vb [hunk 11]
```diff
@@ -26,15 +26,18 @@ Public Class ucrLayerParamsControls
 
     Public Sub SetGeomFunction(clsTempFunction As RFunction)
         clsGeomFunction = clsTempFunction
-        RaiseEvent RParameterChanged(Me)
+        'RaiseEvent RParameterChanged(Me) 'Warning : this was source of problems... the controls should adapt to the content of clsGeomFunction first, then when controls are changed, clsGeomFunction is changed in turn according to the content of the controls.
     End Sub
 
     Public Sub SetLayerParameter(clsTempLayerParam As LayerParameter)
+        'In the following (a little further), the ucrReceiverMetadateProperty will be setup according to the content of clsGeomFunction. While doing so, we don't want to change the content of clsGeomFunction according to the changes that have been made in the ucrReceiverMetadataProperty. Hence we set bEditGeomFunction to False (see ucrLayerParamsControls_RParameterChanged)
```
Comment (FrancoisJRenaud) [inline instat/ucrLayerParamsControls.vb]: Implementing the debug method. Put the bEditGeomFunction = True right at the beginning and = False at very end (of the sub) as when put further inside still got issues. Difficult to track all the RParameterChanged events...
Safer like this and works well when testing.

File: instat/ucrLayerParamsControls.vb [hunk 12]
```diff
@@ -66,7 +70,8 @@ Public Class ucrLayerParamsControls
     End Sub
 
     Private Sub ucrLayerParamsControls_RParameterChanged(ucrControl As ucrLayerParamsControls) Handles Me.RParameterChanged
-        If Not IsNothing(ucrReceiverMetadataProperty.clsLayerParam) AndAlso ucrReceiverMetadataProperty.ctrActive IsNot Nothing AndAlso Not IsNothing(clsGeomFunction) Then
+        'The clsGeomFunction's parameters are adapted to the content of the ucrMetadateReceiver, unless one of the components has not been populated or unless bEditGeomFunction is False.
```
Comment (FrancoisJRenaud) [inline instat/ucrLayerParamsControls.vb]: Explanation + adding the condition bEditGeomFunction before proceding to adapting clsGeomFunction

File: instat/clsRFunction.vb [hunk 13]
```diff
@@ -249,7 +253,7 @@ Public Class RFunction
 
     Public Sub ClearParameters()
         clsParameters.Clear()
-        bIsAssigned = False
+        bIsAssigned = False 'Question to be discussed: Why is the bIsAssigned set to False when Parameters are cleared ?
```
Comment (FrancoisJRenaud) [inline instat/clsRFunction.vb]: Arf yes of course. I m stupid

File: instat/clsRFunction.vb (hunk 10 above)
Comment (FrancoisJRenaud) [inline instat/clsRFunction.vb]: Ok, I think it is not needed. I will just comment it out, test the dialogues again. But things should work fine.

File: instat/ucrGeom.vb (hunk 2 above)
Comment (FrancoisJRenaud) [inline instat/ucrGeom.vb]: sure

File: instat/ucrLayerParameters.vb (hunk 6 above)
Comment (dannyparsons) [inline instat/ucrLayerParameters.vb]: Yes. You don't need to ask about deleting empty subs.

File: instat/ucrLayerParameters.vb (hunk 7 above)
Comment (dannyparsons) [inline instat/ucrLayerParameters.vb]: Yes. You don't need to ask about deleting empty subs.

File: instat/clsRFunction.vb (hunk 10 above)
Comment (FrancoisJRenaud) [inline instat/clsRFunction.vb]: Could add an optional argument: bEraseParametersIfDifferentRCommand with default False to the function ?

Comment (dannyparsons) [inline instat/clsRFunction.vb]: I think it's simpler if not. You may expect the the R command is different and parameters are cleared, if it's not could be confusing. I prefer people do this explicitly.

Comment (FrancoisJRenaud) [inline instat/clsRFunction.vb]: Ok, so I tested with erasing and without. and it is actually necessary to erase... otherwise might run geom_density with geom_boxplot parameters... One thing, dialogue is not remembering settings of other layer when new layer is added, even when same geom. Editing a layer works fine though.

Comment (dannyparsons) [inline instat/clsRFunction.vb]: Can we discuss on skype?

File: instat/ucrGeom.vb [hunk 14]
```diff
@@ -937,11 +937,15 @@ Public Class ucrGeom
...
+
         'Here, the clsCurrGeom is chosen according to the selected geom in cboGeomList. Then the Rcommand of the GeomFunction is chosen accordingly.
         clsCurrGeom = lstAllGeoms(cboGeomList.SelectedIndex)
         'Erase parameters in the clsGeomFunction when geom has indeed changed... necessary to not have irrelevant parameters staying in the geom function when geom has changed...
         If clsGeomFunction.strRCommand <> clsCurrGeom.strGeomName Then
-            clsGeomFunction.ClearParameters()
+            For Each clsParam In lstAllGeoms(lstAllGeoms.FindIndex(Function(x) x.strGeomName = clsGeomFunction.strRCommand)).clsLayerParameters
+                clsGeomFunction.RemoveParameterByName(clsParam.strLayerParameterName)
+            Next
             clsGeomFunction.SetRCommand(clsCurrGeom.strGeomName)
```
Comment (dannyparsons) [inline instat/ucrGeom.vb]: Because the aes are passed in separately they should be dealt with by the aes user control. However, they are also linked to the geom function because they are the mapping parameter of it. You could test how this works for aes parameters. If you edit a layer, it's aes should be filled in on the dialog. Also, if you add a new layer, it should not fill in the aes from the previous layer (but it should fill in global aes). Please test this well, and discuss with @FrancoisJRenaud to make sure it's working.

Comment (FrancoisJRenaud) [inline instat/ucrGeom.vb]: Let me do some testing before merging.

Comment (dannyparsons) [inline instat/ucrGeom.vb]: It should be @maxwellfundi testing as well, so that he only approves it after testing well.

Comment (FrancoisJRenaud) [inline instat/ucrGeom.vb]: yup, will give him my machine when I m done ...

**Comment by FrancoisJRenaud** [comment]
just thinking, the last thing we were worried about is: if add layer with same geom then that layer should be with default layer parameter settings, but that is the case as the GeomFunction is new. There is no communication between layers, apart from what's shared through the ggplot function e g global aesthetics...

**Comment by FrancoisJRenaud** [comment]
@Dannyparsons we have tested this and works well on layer parameters: it is clearing it well when geom is changed;

**Comment by FrancoisJRenaud** [comment]
this was @maxwellfundi
